controllers/controlplane_controller.go

```diff
@@ -48,6 +49,25 @@ func (r *ControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, err
 	}
 
+	// controlplane is deleted, just run garbage collection for cluster wide resources.
+	if !controlplane.DeletionTimestamp.IsZero() {
+		debug(log, "remove owned cluster wide resources", controlplane)
+		var err *multierror.Error
+		removeCrbErr := r.removeOwnedClusterRoleBindings(ctx, controlplane)
+		err = multierror.Append(err, removeCrbErr)
+		removeCrErr := r.removeOwnedClusterRoles(ctx, controlplane)
+		err = multierror.Append(err, removeCrErr)
+		return ctrl.Result{}, err.ErrorOrNil()
+	}
+
+	// ensure the controlplane has a finalizer to delete owned cluster wide resources on delete.
+	metaChanged := k8sutils.EnsureFinalizerInMetadata(&controlplane.ObjectMeta, string(ControlPlaneFinalizerCleanupClusterRole))
+	metaChanged = metaChanged || k8sutils.EnsureFinalizerInMetadata(&controlplane.ObjectMeta, string(ControlPlaneFinalizerCleanupClusterRoleBinding))
+	if metaChanged {
```

Comment:
nit: `metaChanged` is a very general way to describe what the results of the `EnsureFinalizerInMetadata` function does, what do you think about this alternative name?

```suggestion
	finalizersChanged := k8sutils.EnsureFinalizerInMetadata(&controlplane.ObjectMeta, string(ControlPlaneFinalizerCleanupClusterRole))
	finalizersChanged = finalizersChanged || k8sutils.EnsureFinalizerInMetadata(&controlplane.ObjectMeta, string(ControlPlaneFinalizerCleanupClusterRoleBinding))
	if finalizersChanged {
```

Reply:
fixed.